app/services/samples/metadata/update_service.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Samples
+  module Metadata
+    # Service used to Update Samples::Metadata
+    class UpdateService < BaseService
+      SampleMetadataUpdateError = Class.new(StandardError)
+      attr_accessor :sample, :metadata, :analysis_id
+
+      def initialize(project, sample, user = nil, params = {})
+        super(user, params)
+        @project = project
+        @sample = sample
+        @metadata = params['metadata']
+        @analysis_id = params['analysis_id']
+      end
+
+      def execute # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
+        authorize! sample.project, to: :update_sample?
+        metadata_fields_update_status = { updated: [], not_updated: [] }
+
+        if @project.id != @sample.project.id
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.sample_does_not_belong_to_project', sample_name: @sample.name,
+                                                                                      project_name: @project.name)
+        end
+
+        if @metadata.nil? || @metadata == {}
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.empty_metadata', sample_name: @sample.name)
+        else
+          # Without transforming keys, issues with overwritting can occur and multiples of the same key can appear
+          @metadata = @metadata.transform_keys(&:to_s)
+          @metadata.each do |key, value|
+            if value.blank?
+              if @sample['metadata'].key?(key)
+                @sample['metadata'].delete(key)
+                @sample['metadata_provenance'].delete(key)
+                metadata_fields_update_status[:updated].append(key)
+              end
+            else
+              metadata_updated = update_metadata(key, value)
+              if metadata_updated
+                metadata_fields_update_status[:updated].append(key)
+              else
+                metadata_fields_update_status[:not_updated].append(key)
+              end
+            end
+          end
+          @sample.update(id: @sample.id)
+
+          if metadata_fields_update_status[:not_updated].count.positive?
+            raise SampleMetadataUpdateError,
+                  I18n.t('services.samples.metadata.user_cannot_update_metadata',
+                         sample_name: @sample.name,
+                         metadata_fields: metadata_fields_update_status[:not_updated].join(', '))
+          else
+            metadata_fields_update_status[:updated]
+          end
+
+        end
+      rescue Samples::Metadata::UpdateService::SampleMetadataUpdateError => e
+        @sample.errors.add(:base, e.message)
+        metadata_fields_update_status[:updated]
```

Comment:
**nitpick (blocking):** Should this have `:updated` status when there is an error?

Reply:
I discussed this above with Deep https://github.com/phac-nml/irida-next/pull/311#discussion_r1417766067

This is mainly to capture if metadata was passed and some fields updated, while others did not (mainly by if user tries to update a field already updated by an analysis). This is under the assumption we want to provide the user with both a success and failure feedback for the fields that were updated and not updated via the controller. Without returning the `updated` list, I couldn't think of a way to check if some fields were updated. So this way, we can check if any fields were updated with the returned list, and anything `not_updated` would be contained in the `@sample.errors`.

Would love to hear if you have another idea on how to approach this solution if one comes to mind!

Reply:
This has been updated to just return the `metadata_fields_update_status` which is a hash that includes updated: [] and not_updated: []. In the controller we can just check to see if `updated` has a length to determine whether it will go into the success block or the error block of the if/else in the controller